Add openebs-mixin

rajaSahil opened the pull request:
Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

### Reviewer Notes

- Please ignore the changes in [monitoring/jsonnet/manifests/](https://github.com/rajaSahil/monitoring/tree/Add-Openebs-Mixin/jsonnet/manifests) directory. These are the files generated through the help of jsonnet and gojsontoyaml.

- Please ignore the changes in [monitoring/deploy/charts/openebs-monitoring/dashboards](https://github.com/rajaSahil/monitoring/tree/Add-Openebs-Mixin/deploy/charts/openebs-monitoring/dashboards) and [monitoring/deploy/charts/openebs-monitoring/rules](https://github.com/rajaSahil/monitoring/tree/Add-Openebs-Mixin/deploy/charts/openebs-monitoring/rules) as these are generated using jsonnet. Commit: 6ef162eba591f54ebc5442232a64aaa84390e9b0

- All the dashboards json present in [monitoring/jsonnet/openebs-mixin/dashboards](https://github.com/rajaSahil/monitoring/tree/Add-Openebs-Mixin/jsonnet/openebs-mixin/dashboards) are same as that of [dashboards](https://github.com/openebs/monitoring/tree/develop/deploy/charts/openebs-monitoring/dashboards).

- Alert rules defined in [monitoring/jsonnet/openebs-mixin/rules/](https://github.com/rajaSahil/monitoring/tree/Add-Openebs-Mixin/jsonnet/openebs-mixin/rules) are taken from [rules](https://github.com/openebs/monitoring/tree/develop/deploy/charts/openebs-monitoring/rules)

### What this PR does?

- Add openebs-mixin
- Add openebs monitoring using jsonnet
- Generate dashboards and alert rules for helm charts through openebs-mixin.

#### openebs-mixin
- Consists of all the dashboards and alert rules for openebs.

#### openebs monitoring using jsonnet
- openebs monitoring using jsonnet which has an external dependency of kube-Prometheus. It uses openebs-mixin for dashboards and alert rules.

#### Directory structure:
```
openebs/monitoring/jsonnet
├── addons # openebs-monitoring addon
│   ├── dashboard-cm.libsonnet # dashboard configmap to include openebs-* dashboard json files from openebs-mixin
│   ├── podMonitors.libsonnet # podMonitor to be created for different cas types
│   ├── prometheusRules.libsonnet # create prometheusRule object which includes alert rules defined in openebs-mixin
│   └── serviceMonitors.libsonnet # serviceMonitor to be created for different cas types
├── buildscripts # scripts and code required to build the project
│   ├── build.sh # script to generate manifests
│   ├── go.mod
│   ├── go.sum
│   └── tools.go
├── config.libsonnet # configuration to be used for openebs monitoing
├── jsonnetfile.json # direct dependencies
├── jsonnetfile.lock.json # all dependencies with exact version
├── lib # contains code that is used in different places
│   └── utils.libsonnet # code which is used to create openebs-monitoring addons
├── main.jsonnet # entrypoint of code
├── Makefile
├── manifests # generated yamls
│   ├── openebs-addons # openebs-monitoring addons yaml
│   │   ├── openebs-*.yaml
│   ├── setup # kube-prometheus crds, rbac and prometheus-operator yaml
│   │   ├── ...
│   ├── ... # kube-prometheus components yaml
└── openebs-mixin # openebs-mixin for dashboards and alert rules
    ├── build.sh
    ├── config.libsonnet # configuration that is used while writing dashboards and alert rules
    ├── dashboards
    │   ├── cstor-overview.json
    │   ├── cstor-pool.json
    │   ├── cstor-volume.json
    │   ├── cstor-volume-replica.json
    │   ├── dashboards.libsonnet # to import all the dashboard json
    │   ├── jiva-volume.json
    │   └── localpv-workload.json
    ├── jsonnetfile.json # direct dependencies
    ├── lib
    │   ├── dashboards.jsonnet
    │   └── rules.jsonnet
    ├── Makefile
    ├── mixin.libsonnet # openebs-mixin that imports all the dashboards and alert rules
    └── rules # contains alert rules for different cas types
        ├── cstor-pool-rules.libsonnet
        ├── cstor-rules.libsonnet
        ├── cstor-volume-rules.libsonnet
        ├── jiva-rules.libsonnet
        ├── jiva-volume-rules.libsonnet
        ├── rules.libsonnet
        └── volume-rules.libsonnet
```

#### Note: 
 Refer to this document for more insight into openebs-mixin and openebs-monitoring using jsonnet: 

- https://docs.google.com/document/d/1xJAjByPysQUHkxGK8gNUAALTt3f-eggGdLpNX4PiXn4/edit?usp=sharing
- https://docs.google.com/document/d/10SDX5dAQIoM5C3TP4QM4RibefVpSPXDy7a-NNUohjX0/edit?usp=sharing

rajaSahil committed:
Add openebs-mixin

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Added openebs-mixin

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Formatted jsonnet

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Generated dashboards and rules json

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Updated values yaml and prometheusrule template

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Updated gitignore

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

rajaSahil committed:
Add openebs-monitoring in jsonnet

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Added comments and updated config

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Bumped chart version

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

generated manifests

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Minor fix

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Generated cstor prometheusRules

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Added configuration for device-localPV and configuration for openebs-mixin

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Fix variable name

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Added README

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Update README

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Added volume-node rules in openebs-mixin

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Bumped chart version

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

rajaSahil commented on deploy/charts/openebs-monitoring/templates/prometheusRules.yaml:
```diff
@@ -5,12 +5,12 @@
 {{- end }}
 {{- end }}
-{{- $files := .Files.Glob "rules/**.yaml" }}
+{{- $files := .Files.Glob "rules/**.json" }}
```
Alert rules will be generated in json using jsonnet.

rajaSahil commented on deploy/charts/openebs-monitoring/values.yaml:
```diff
@@ -178,7 +178,7 @@ alertRules:
   Jiva:
     enabled: true
-  volumes:
+  Volume:
```
volume-rules.json is present in `rules/Volume/`.

avishnu commented on jsonnet/buildscripts/build.sh:
```diff
@@ -0,0 +1,22 @@
+jsonnet -J vendor -m manifests "${1-main.jsonnet}" | xargs -I{} sh -c 'cat {} | gojsontoyaml > {}.yaml' -- {}
+
+# Make sure to remove json files
+find manifests -type f ! -name '*.yaml' -delete
```
what is ! for?

avishnu replied:
Is it to exclude yaml files?

rajaSahil replied:
> Is it to exclude yaml files?

Yes

avishnu commented on jsonnet/config.libsonnet:
```diff
@@ -0,0 +1,237 @@
+        // To generate manifests for cstor. If set, manifests will be generated for cStor.
+        enabled: true,
+        // To generate dashboards configMap yamls. If set, dashboards will be appended in grafana-dashboardDefinition yaml.
+        installDashboards: true,
```
should this be just dashboards as below there is just alertRules?

avishnu commented on jsonnet/config.libsonnet:
```diff
@@ -0,0 +1,237 @@
+          enabled: false,
+        },
+      },
+      lvmLocalPV: {
```
Can we add one for device localpv as well?

rajaSahil replied:
Yes, will add it.

avishnu commented on jsonnet/jsonnetfile.json:
```diff
@@ -0,0 +1,15 @@
+{
```
Is the filename as per the standard or can it be dependency-jsonnet.json?

rajaSahil replied:
It is the auto-generated file when we run `jb init`.

avishnu commented on jsonnet/openebs-mixin/rules/prometheusRules.libsonnet:
```diff
@@ -0,0 +1,11 @@
+local cstor = (import 'cstor-rules.libsonnet');
```
Bit inconsistent on naming the file, cstor-rules and jiva-rules, but prometheusRules.

rajaSahil replied:
Changed it to `prometheus-rules`

rajaSahil committed:
Changed volume-node rules to npd-rules

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Removed manifests

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Removed generated dashboards and rules for helm

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Update separator for relabelings

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Add rules folder for NPD

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Minor fix

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

rajaSahil committed:
Generated dashboards and rules for helm charts

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

avishnu commented on jsonnet/openebs-mixin/build.sh:
```diff
@@ -26,7 +26,7 @@ generateRules(){
 	mkdir -p  $rulesDirPath
 	jsonnet -J vendor -m $rulesDirPath lib/rules.jsonnet
-	rulesFolder=(cStor Jiva Volume)
+	rulesFolder=(cStor Jiva Volume NPD)
```
Is there an extra space between cStor and Jiva?

rajaSahil replied:
No, it's fine. There is only one space present.

avishnu commented on jsonnet/openebs-mixin/rules/npd-rules.libsonnet:
```diff
@@ -1,8 +1,8 @@
 function(param) {
-  local rules = self,
+  //local rules = self,
```
Why is this commented?

rajaSahil replied:
For now, we are not using any configuration in defining the rules. That's why it is not required.

avishnu commented on jsonnet/openebs-mixin/rules/volume-rules.libsonnet:
```diff
@@ -1,12 +1,11 @@
   prometheusRules+:: {
     volume: {
-      [if rules._config.alertRules.volume then 'groups']+: volumePVC(rules._config).prometheusRules.volumePVC.groups + volumeNode(rules._config).prometheusRules.volumeNode.groups,
+      [if rules._config.alertRules.volume then 'groups']+: volumePVC(rules._config).prometheusRules.volumePVC.groups,
```
Why this change?

rajaSahil replied:
volume-node-rules is changed to npd-rules.

avishnu commented on deploy/charts/openebs-monitoring/rules/cStor/cstor-rules.json:
```diff
@@ -0,0 +1,44 @@
+                  "storageClassClaim": "{{ $labels.storage_pool_claim }}",
+                  "summary": "cStor pool '{{$labels.cstor_pool}}' created for pool claim '{{$labels.storage_pool_claim}}' has {{ with printf \"openebs_free_pool_capacity{cstor_pool='%s',instance='%s',job='%s',kubernetes_pod_name='%s',slave='%s',storage_pool_claim='%s'}/(1024*1024*1024)\" $labels.cstor_pool $labels.instance $labels.job $labels.kubernetes_pod_name $labels.slave $labels.storage_pool ...
+               },
+               "expr": "openebs_used_pool_capacity_percent > 80",
```
These numeric values should be made configurable. Can be taken up in future PR.

rajaSahil replied:
Sure.

avishnu commented on jsonnet/README.md:
````diff
@@ -0,0 +1,36 @@
+	$ make generate
+	```
+
+The  files in `manifests/` are the yamls that has to be applied on the kubernetes cluster.
````
have* to be applied

avishnu commented on jsonnet/README.md:
```diff
@@ -0,0 +1,36 @@
+The  files in `manifests/` are the yamls that has to be applied on the kubernetes cluster.
+
+### Apply the kube-prometheus stack and openebs monitoring addons
+The previous steps (Generate manifests) has created a bunch of manifest files in the manifests/ folder. Now simply use kubectl to install Prometheus and Grafana as per your configuration:
```
Should be step-3?

rajaSahil replied:
Yes.

avishnu commented on jsonnet/config.libsonnet:
```diff
@@ -0,0 +1,258 @@
+  },
+
+  // Template function for PodMonitor
+  local podMonitor(port, path, selectorName) = {
```
where are we using podMonitors?

rajaSahil replied:
For cStor pool pods, we are creating a PodMonitor.

avishnu commented on jsonnet/openebs-mixin/README.md:
````diff
@@ -0,0 +1,62 @@
+1. Install `jsonnet` and `jb`
+	```
+	$ go get github.com/jsonnet-bundler/jsonnet-bundler/cmd/jb
+	$ brew install jsonnet
````
How is brew installed?

rajaSahil replied:
Have to add `sudo apt-get install jsonnet` for linux.
brew is for mac.

avishnu replied:
Add for linux.

avishnu commented on jsonnet/openebs-mixin/README.md:
````diff
@@ -0,0 +1,62 @@
+	$ make generate
+	```
+
+The  files in `rulesDirPath` are the alert rules defined for openebs which has to be used to create prometheusRule yaml then need to passed to your Prometheus server, and the files in `dashboardsDirPath` need to be imported into you Grafana server. 
````
need to be imported into your Prometheus.
need to be imported into your Grafana.

avishnu replied:
The files (remove extra space between The and files) in `rulesDirPath` contain Prometheus-based alert rules for OpenEBS. These files will be aggregated to form a prometheusRule yaml that need to be imported into your Prometheus instance.
Similarly, the files in `dashboardsDirPath` are dashboard json files that need to be imported into your Grafana instance.

avishnu commented on jsonnet/openebs-mixin/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## Configuration
+
+The available fields and their default values can be seen in `config.libsonnet`. 
```
are present in

avishnu commented on jsonnet/openebs-mixin/README.md:
```diff
@@ -0,0 +1,62 @@
+    alertRules: {
+      cStor: true,
+      Jiva: true,
+      lvmLocalPV: false,
```
LVMLocalPV

avishnu commented on jsonnet/openebs-mixin/README.md:
```diff
@@ -0,0 +1,62 @@
+      cStor: true,
+      Jiva: true,
+      lvmLocalPV: false,
+      volume: true,
```
How about AllVolumeTypes?

avishnu commented on jsonnet/openebs-mixin/README.md:
```diff
@@ -0,0 +1,62 @@
+      cStor: true,
+      Jiva: true,
+      lvmLocalPV: false,
+      volume: true,
```
Since this category `volume` contains common rules for all types of CAS engines, will it be better if we remove the switch for this and include the rules whenever any of the above CAS types are set as true?
The contention is, why would anyone want to enable cStor=true, but volume=false?

avishnu replied:
Should we specify npd as well here?

rajaSahil replied:
Makes sense.
We can do an OR operation to check whether any of the CAS types are set as true.
 `volume: $._config.alertRules.cStor || $._config.alertRules.Jiva || $._config.alertRules.lvmLocalPV`

avishnu replied:
Ok, make it a non-configurable variable then.

Ab-hishek replied:
This configuration section is not in sync with the actual configuration.

avishnu commented on jsonnet/openebs-mixin/build.sh:
```diff
@@ -0,0 +1,40 @@
+	do
+	    x=`echo $i | awk '{print tolower($0)}'`
+		mkdir -p $rulesDirPath/$i
+		find ./$rulesDirPath  -name "*$x*" |xargs -I{} sh -c "mv {} $rulesDirPath/$i/"
```
Is there an extra space or two before -name?

avishnu commented on jsonnet/openebs-mixin/config.libsonnet:
```diff
@@ -0,0 +1,19 @@
+      npd: true,
+    },
+  },
+}
```
Can we make everything in lower-case to avoid any confusion?

rajaSahil replied:
We can but it might create confusion for lvmlocalpv or devicelocalpv.

avishnu replied:
I don't think there will be any confusion.

rajaSahil replied:
Okay. Will update it.

avishnu commented on jsonnet/openebs-mixin/rules/jiva-volume-rules.libsonnet:
```diff
@@ -0,0 +1,30 @@
+function(param) {
```
Basic question: can we combine jiva-rules.libsonnet and jiva-volume-rules.libsonnet into one?

rajaSahil replied:
Yes, we can.

rajaSahil committed:
Addressed review comments

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Addressed review comments

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Addressed review comments

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

fixed linting error

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

avishnu commented on jsonnet/openebs-mixin/config.libsonnet:
```diff
@@ -1,18 +1,25 @@
     alertRules: {
-      cStor: true,
-      Jiva: true,
-      lvmLocalPV: false,
```
Is this not needed?

rajaSahil replied:
No, we don't have any alert rules for lvm-localPV.

avishnu commented on jsonnet/openebs-mixin/build.sh:
```diff
@@ -18,20 +18,20 @@ generateDashboards(){
 	    x=`echo $i | awk '{print tolower($0)}'`
 		mkdir -p $rulesDirPath/$i
-		find ./$rulesDirPath  -name "*$x*" |xargs -I{} sh -c "mv {} $rulesDirPath/$i/"
+		find ./$rulesDirPath -type f -name "*$x*" |xargs -I{} sh -c "mv {} $rulesDirPath/$i/" 
```
Extra space at the end of line

avishnu commented on jsonnet/openebs-mixin/rules/rules.libsonnet:
```diff
@@ -1,9 +1,25 @@
+local volume = (import './volume/volume-rules.libsonnet');
+local npd = (import './npd/npd-rules.libsonnet');
+
+// Alert rules for different CAS types and others(as defined in config.libsonnet under alertRules) are stored seperately in prometheusRules object.
```
separately*

rajaSahil committed:
Added github workflow for jsonnet formatter and linter

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

rajaSahil committed:
Dashboard changes for cstor and localpv

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

avishnu reviewed (commented):
Lgtm

avishnu commented on .github/workflows/jsonnet-lint.yml:
```diff
@@ -0,0 +1,30 @@
+
+      - name: Jsonnet linter
+        run: |
+          cd jsonnet && make --always-make lint
```
Is this tested?

rajaSahil replied:
Yes. https://github.com/openebs/monitoring/pull/39/checks?check_run_id=3193178661

rajaSahil committed:
Added npd dashboard to openebs-mixin

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Generated npd dashboard for helm charts through mixin

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Generated dashboard

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

avishnu reviewed (approved):
lgtm

rajaSahil committed:
Move lvm dashboards and alerts to mixin

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

rajaSahil committed:
Generate dashboard and alert rules for lvm localpv

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

rajaSahil committed:
Updated app version

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

Ab-hishek reviewed (approved):
On small minor comment else everything else looks good.

rajaSahil committed:
Updated readme

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

rajaSahil committed:
Update dashboards

Signed-off-by: Sahil Raja <sahilraja242@gmail.com>

avishnu commented on jsonnet/openebs-mixin/dashboards/openebs/localpv-workload.json:
```diff
@@ -1,4216 +1,4195 @@
+      "1d"
+    ],
+    "time_options": ["5m", "15m", "1h", "6h", "12h", "24h", "2d", "7d", "30d"]
+  },
```
What are the changes exactly in this commit?

rajaSahil replied:
cStor dashboards were updated for helm charts which I forgot to update for mixin so updated the cStor dashboards and few formatting changes.
